### PR TITLE
feat: Minecraft 1.21.4

### PR DESCRIPTION
**Pull request by ccetl**
Ports the client to 1.21.4.

TODOs:
- [x] finish reimplementing the blur ui shader (-> refactor overlay sampler to uniformI)
   - [ ] fix weird nametag text rendering when spoofing the main framebuffer
-  Esp
   - [x] fix the "glow" making entities semi transparent
   - [x] finish the reimplementation of the outline mode
- [ ] calculate the damage dealt to the fakeplayer correctly
- [x] check if using bundles slow you down (and therefore should be included in NoSlow)
- [x] Framebuffershaders and UI blur break the HUD when used without Nametags enabled

Non-Goals (should be implemented separately):
- take the 1.21.4 tnt minecard additional parameters to calculate the explosion damage into account
- send the tick end packet after every move packet
- fully fix the fakeplayer

The code to render the browser image is taken from #4452.

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -20,29 +20,29 @@ kotlin.code.style=official
...
 mod_version=0.22.1
 maven_group=net.ccbluex
 archives_base_name=liquidbounce
 # Kotlin
 # https://kotlinlang.org/releases.html
-kotlin_version=2.1.0
+kotlin_version=2.0.21
 # https://maven.fabricmc.net/net/fabricmc/fabric-language-kotlin/
-fabric_kotlin_version=1.13.0+kotlin.2.1.0
+fabric_kotlin_version=1.12.3+kotlin.2.0.21
```
Comment (DataM0del) [inline gradle.properties]: Why are we downgrading Kotlin?

File: gradle.properties [hunk 2]
```diff
@@ -20,29 +20,29 @@ kotlin.code.style=official
...
 # mcef
-mcef_version=1.3.1-1.21.1
+mcef_version=1.3.0-1.21.4
```
Comment (DataM0del) [inline gradle.properties:42]: Waiting for the day that @1zun4 will release MCEF 1.3.1 for 1.21.4

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/network/MixinClientPlayNetworkHandler.java [hunk 3]
```diff
@@ -82,27 +78,6 @@ private void onChunkDeltaUpdateEnd(ChunkDeltaUpdateS2CPacket packet, CallbackInf
...
-            if (fixedX != x || fixedY != y || fixedZ != z) {
-                ModuleAntiExploit.INSTANCE.notifyAboutExploit("Limited too strong explosion",
-                        true);
-                args.set(0, fixedX);
-                args.set(1, fixedY);
-                args.set(2, fixedZ);
-            }
-        }
-    }
-
```
Comment (DataM0del) [inline src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/network/MixinClientPlayNetworkHandler.java]: Not sure if mine actually worked, but at least try to do something?

**Comment by DataM0del** [comment]
I'm going to probably now work on this branch instead of mine, since this is better than mine and I closed it anyways (because no stupid rendering offset momento), and would probably Forever Detekted™.

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/render/MixinEntityRenderer.java [hunk 4]
```diff
@@ -98,11 +99,17 @@ private void renderLabel(Entity entity, OrderedText text, MatrixStack matrices,
...
         if (ModuleNametags.INSTANCE.getRunning() && ModuleNametags.shouldRenderNametag(entity)) {
             ci.cancel();
         }
     }
 
+    @Inject(method = "updateRenderState", at = @At("HEAD"))
+    private void hookInjectEntityIntoState(T entity, S state, float tickDelta, CallbackInfo ci) {
+        ((EntityRenderStateAddition) state).liquid_bounce$setEntity(entity);
+    }
+
```
Comment (DataM0del) [inline src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/render/MixinEntityRenderer.java:114]: POV: liquidskid if this was his code being ported to legacy
> oh no my code being ported to a different branch of the same project stop it now 2024 now lets paste everything from legacy and act like its fine and we're not contradicting our own thing

**Review by DataM0del — CHANGES_REQUESTED** [review]
please no downgrading of Kotlin :pray:

**Comment by DataM0del** [comment]
> fix the "glow" making entities semi transparent

I think that's because I forgot to invert a boolean or something, it was fixed in 91d2b5831743e21de02b715e7996f3a33f86eab6

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/render/MixinLivingEntityRenderer.java [hunk 5]
```diff
@@ -65,22 +69,22 @@ private void injectRender(T livingEntity, float f, float g, MatrixStack matrixSt
...
-    private boolean injectTrueSight(boolean original, T livingEntity) {
+    @ModifyExpressionValue(method = "render(Lnet/minecraft/client/render/entity/state/LivingEntityRenderState;Lnet/minecraft/client/util/math/MatrixStack;Lnet/minecraft/client/render/VertexConsumerProvider;I)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/render/entity/LivingEntityRenderer;isVisible(Lnet/minecraft/client/render/entity/state/LivingEntityRenderState;)Z"))
+    private boolean injectTrueSight(boolean original, @Local(argsOnly = true) S livingEntityRenderState) {
         // Check if TrueSight is enabled and entities are enabled or ESP is enabled and in glow mode
         if (ModuleTrueSight.INSTANCE.getRunning() && ModuleTrueSight.INSTANCE.getEntities() ||
-                ModuleESP.INSTANCE.getRunning() && ModuleESP.INSTANCE.requiresTrueSight(livingEntity)) {
+                ModuleESP.INSTANCE.getRunning() && ModuleESP.INSTANCE.requiresTrueSight(((LivingEntity) ((EntityRenderStateAddition) livingEntityRenderState).liquid_bounce$getEntity()))) {
             return false;
```
Comment (DataM0del) [inline src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/render/MixinLivingEntityRenderer.java]:
```suggestion
            return true;
```
from 91d2b5831743e21de02b715e7996f3a33f86eab6

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/network/MixinClientPlayNetworkHandler.java (hunk 3 above)
Comment (ccetl) [reply src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/network/MixinClientPlayNetworkHandler.java]: Minecraft doesn't calculate explosions client side anymore, this exploit is fixed.

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/render/MixinEntityRenderer.java (hunk 4 above)
Comment (ccetl) [reply src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/render/MixinEntityRenderer.java:114]: ???

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/network/MixinClientPlayNetworkHandler.java (hunk 3 above)
Comment (DataM0del) [reply src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/network/MixinClientPlayNetworkHandler.java]: Oh, Ok. not sure about if the server sends some big knockback or something (if that's covered in anti exploit) and the client takes it and crashes it.

File: gradle.properties (hunk 1 above)
Comment (ccetl) [reply gradle.properties]: I'll take care of the versions as the last step.

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/render/MixinLivingEntityRenderer.java (hunk 5 above)
Comment (ccetl) [reply src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/render/MixinLivingEntityRenderer.java]: Thank you.

File: gradle.properties (hunk 2 above)
Comment (1zun4) [reply gradle.properties:42]: You know that it does not matter?

Comment (DataM0del) [reply gradle.properties:42]: > You know that it does not matter?

Yes, but I should probably fork MCEF, pull the upstream changes from 1.21.1, and get ccetl to use mine like I did in my PR. Or you could just click the merge button in the branch and release it.

**Comment by DataM0del** [comment]
It seems like the ExplodeBypass exploit on loyisa crashes LiquidBounce successfully (it doesn't instantly crash and close though).
image
image

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/network/MixinClientPlayNetworkHandler.java (hunk 3 above)
Comment (DataM0del) [reply src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/network/MixinClientPlayNetworkHandler.java]: > Minecraft doesn't calculate explosions client side anymore, this exploit is fixed.

It doesn't calculate explosions client side anymore, but what does the client do? It handles the knockback.
And ExplodeBypass crashes without this, also, my implementation works.

> Oh, Ok. not sure about if the server sends some big knockback or something (if that's covered in anti exploit) and the client takes it and crashes it.

And this was true, I guess.

**Comment by DataM0del** [comment]
detekted

**Comment by DataM0del** [comment]
undetekted

File: gradle.properties (hunk 2 above)
Comment (1zun4) [reply gradle.properties:42]: No, we will not use "yours" just because you want it to use a version that does not change anything relevant for 1.21.4

Comment (1zun4) [reply gradle.properties:42]: Here as comparison (1.21.1): https://github.com/CCBlueX/mcef/compare/1.3.0-1.21.1...1.3.1-1.21.1
The release of 1.3.1 was to bump dependencies on 1.21.1 - which was ALREADY done on 1.21.4 branch.

https://github.com/CCBlueX/mcef/blob/1.21.4/gradle.properties

**Review by DataM0del — APPROVED** [review]

**Comment by 1zun4** [comment]
1. Looks like the /itemTexture endpoint is not working.
image


2. Blending mode for browser texture is wrong. That is caused by the blur.
image

For the browser drawing, we use the following blending - but when using the HUD Blur this seem to be overwritten.
image

This can be fixed by changing the HUD Blur blending:
```patch
Index: src/main/kotlin/net/ccbluex/liquidbounce/render/engine/UiRenderer.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/kotlin/net/ccbluex/liquidbounce/render/engine/UiRenderer.kt b/src/main/kotlin/net/ccbluex/liquidbounce/render/engine/UiRenderer.kt
--- a/src/main/kotlin/net/ccbluex/liquidbounce/render/engine/UiRenderer.kt	(revision 4e50c8d699998e89010330059ce3f6456c33341c)
+++ b/src/main/kotlin/net/ccbluex/liquidbounce/render/engine/UiRenderer.kt	(date 1734939970264)
@@ -108,7 +108,7 @@
         blur()
 
         RenderSystem.enableBlend()
-        RenderSystem.blendFunc(GlStateManager.SrcFactor.SRC_ALPHA, GlStateManager.DstFactor.ONE_MINUS_SRC_ALPHA)
+        RenderSystem.blendFunc(GlStateManager.SrcFactor.ONE, GlStateManager.DstFactor.ONE_MINUS_SRC_ALPHA)
 
         this.overlayFramebuffer.drawInternal(mc.window.framebufferWidth, mc.window.framebufferHeight)
 
```

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/client/ModuleRichPresence.kt [hunk 6]
```diff
@@ -61,7 +61,7 @@ object ModuleRichPresence : ClientModule("RichPresence", Category.CLIENT, state
     aliases = arrayOf("DiscordPresence")) {
 
     private val detailsText by text("Details", "Nextgen v%clientVersion% by %clientAuthor%")
-    private val stateText by text("State", "%enabledModules% of %totalModules% modules enabled")
+    private val stateText by text("MixinEntityRenderState", "%enabledModules% of %totalModules% modules enabled")
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/client/ModuleRichPresence.kt]: image

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/movement/noslow/ModuleNoSlow.kt [hunk 7]
```diff
@@ -56,28 +57,18 @@ object ModuleNoSlow : ClientModule("NoSlow", Category.MOVEMENT) {
...
-
-        UseAction.BOW, UseAction.CROSSBOW, UseAction.SPEAR -> if (NoSlowBow.enabled) Pair(
-            NoSlowBow.forwardMultiplier, NoSlowBow.sidewaysMultiplier
-        ) else Pair(0.2f, 0.2f)
-
+        UseAction.NONE -> NoSlowUseActionHandler.DEFAULT_USE_MUL
+        UseAction.EAT, UseAction.DRINK -> NoSlowConsume.getMultiplier()
+        UseAction.BLOCK, UseAction.SPYGLASS, UseAction.TOOT_HORN, UseAction.BRUSH -> NoSlowBlock.getMultiplier()
+        UseAction.BOW, UseAction.CROSSBOW, UseAction.SPEAR -> NoSlowBow.getMultiplier()
+        UseAction.BUNDLE -> NoSlowBundle.getMultiplier()
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/movement/noslow/ModuleNoSlow.kt:71]: 👍🏻

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/ModuleCustomAmbience.kt [hunk 8]
```diff
@@ -57,42 +62,35 @@ object ModuleCustomAmbience : ClientModule("CustomAmbience", Category.RENDER) {
         /**
          * [MixinBackgroundRenderer]
          */
-        fun modifyFog(camera: Camera, viewDistance: Float) {
+        fun modifyFog(camera: Camera, viewDistance: Float, fog: Fog): Fog {
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/ModuleCustomAmbience.kt:65]: 👍🏻

File: src/main/kotlin/net/ccbluex/liquidbounce/utils/block/BlockExtensions.kt [hunk 9]
```diff
@@ -489,7 +489,8 @@ private inline fun handleActionsOnAccept(
     onPlacementSuccess: () -> Boolean,
     swingMode: SwingMode = SwingMode.DO_NOT_HIDE,
 ) {
-    if (!interactionResult.shouldSwingHand()) {
+    if (interactionResult !is ActionResult.Success ||
+        interactionResult.swingSource != ActionResult.SwingSource.SERVER) {
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/utils/block/BlockExtensions.kt]: this should be an extension

File: src/main/resources/liquidbounce.mixins.json [hunk 10]
```diff
@@ -104,5 +105,9 @@
   "server": [],
   "injectors": {
     "defaultRequire": 1
-  }
+  },
+  "mixins": [
+    "minecraft.item.MixinArmorItem",
+    "minecraft.item.MixinMiningToolItem"
+  ]
```
Comment (superblaubeere27) [inline src/main/resources/liquidbounce.mixins.json]: This should be in the client section.

**Review by superblaubeere27 — CHANGES_REQUESTED** [review]

**Review by superblaubeere27 — APPROVED** [review]

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/client/ModuleRichPresence.kt (hunk 6 above)
Comment (DataM0del) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/client/ModuleRichPresence.kt]: :fire:

Comment (ccetl) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/client/ModuleRichPresence.kt]: xD

**Comment by 1zun4** [comment]
The PR was not ready at all.

**Comment by 1zun4** [comment]
I will force push the main branch back to the original, please make sure to direct PRs towards 1.21.4 to: https://github.com/CCBlueX/LiquidBounce/commits/nextgen-1.21.4 for now.
It also seems like the LiquidLauncher is not ready for Minecraft 1.21.4 yet, which is also another big factor that makes it impossible to fix this issue right now, including the following issues:
- Missing compression dependency (https://github.com/CCBlueX/LiquidBounce/issues/5037)
- Browser Drawer accessing protected field (https://paste.ccbluex.net/ada047c4e)

**Comment by ccetl** [comment]
> The PR was not ready at all.

? Like what

**Comment by 1zun4** [comment]
> > The PR was not ready at all.
> 
> ? Like what

It was not tested on the LiquidLauncher and other launchers  You can't expect something to run without testing it...

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/render/MixinLivingEntityRenderer.java (hunk 5 above)
Comment (ccetl) [reply src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/render/MixinLivingEntityRenderer.java]: This is not fixed with that.

**Comment by ccetl** [comment]
> > fix the "glow" making entities semi transparent
> 
> I think that's because I forgot to invert a boolean or something, it was fixed in [91d2b58](https://github.com/CCBlueX/LiquidBounce/commit/91d2b5831743e21de02b715e7996f3a33f86eab6)

image

Some entities are still semi transparent.

**Pull request merged**